compiler/src/dotty/tools/dotc/util/Signatures.scala

```diff
@@ -41,109 +43,238 @@ object Signatures {
-      // This is necessary to extract proper result type as unapply can return other methods eg. apply
-      case MethodTpe(_, _, resultType) =>
-        extractParamTypess(resultType.widenDealias)
-      case _ =>
-        Nil
-    }
+    tree.symbol != NoSymbol && ctx.definitions.isTupleClass(tree.symbol.owner.companionClass)
+
+  /**
+   * Get unapply method result type omiting unknown types and another method calls.
+   *
+   * @param fun Unapply tree
+   * @return Proper unapply method type after extracting result from method types and omiting unknown types.
+   */
+  private def unapplyMethodResult(fun: tpd.Tree)(using Context): Type =
+    val typeWithoutBinds = fun match
+      case TypeApply(_, Bind(_, _) :: _) => fun.symbol.asSingleDenotation.info
+      case other => other.tpe
+
+    typeWithoutBinds.finalResultType.widenDealias match
+      case methodType: MethodType => methodType.resultType.widen
+      case other => other
 
-  // Returns extractors from given type. In case if there are no extractor methods it fallbacks to get method
-  private def getExtractorMembers(resultType: Type)(using Context): List[List[Type]] =
+  /**
+   *  Maps type by checking if given match is single match, name-based match, sequence match or product sequence match.
+   *  The precedence in higher priority - higher index order for fixed-arity extractors is:
+   *    1. Single match
+   *    2. Name based match
+   *  For variadic extractors:
+   *    1. Sequence match
+   *    2. Product sequence match
+   *
+   *  @see [[https://docs.scala-lang.org/scala3/reference/changed-features/pattern-matching.html]]
+   *
+   *  @param resultType Final result type for unapply
+   *  @param patternCount Currently applied patterns to unapply function
+   *  @param isUnapplySeq true if unapply name equals "unapplySeq", false otherwise
+   *  @return List of List of types dependent on option less extractor type.
+   */
+  private def mapOptionLessUnapply(
+    resultType: Type,
+    patternCount: Int,
+    isUnapplySeq: Boolean
+  )(using Context): List[List[Type]] =
     val productAccessors = resultType.memberDenots(
       underscoreMembersFilter,
       (name, buf) => buf += resultType.member(name).asSingleDenotation
     )
-    val availableExtractors = if productAccessors.isEmpty then
-      List(resultType.member(core.Names.termName("get")))
-    else
-      productAccessors
-    List(availableExtractors.map(_.info.finalResultType.stripAnnots).toList)
 
-  object underscoreMembersFilter extends NameFilter {
-    def apply(pre: Type, name: Name)(using Context): Boolean = name.startsWith("_")
+    val getMethod = resultType.member(core.Names.termName("get"))
+    val dropMethod = resultType.member(core.Names.termName("drop"))
+
+    val availableExtractors = (getMethod.exists && patternCount <= 1, isUnapplySeq && dropMethod.exists) match
+      case (_, true) => List(dropMethod)
+      case (true, _) => List(getMethod)
+      case _  => productAccessors
```

Comment:
Just an if seems more useful, not separating the conditions from the results
```suggestion
    if isUnapplySeq && dropMethod.exists then
      List(dropMethod)
    else if getMethod.exists && patternCount <= 1 then
      List(getMethod)
    else
      productAccessors
```